pyro/optim/multi.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+
+from pyro.ops.newton import newton_step_2d
+from pyro.optim.optim import PyroOptim
+
+
+class MultiOptimizer(object):
+    """
+    Base class of optimizers that make use of higher-order derivatives.
+
+    Higher-order optimizers generally use :func:`torch.autograd.grad` rather
+    than :meth:`torch.Tensor.backward`, and therefore require a different
+    interface from usual Pyro and PyTorch optimizers. In this interface,
+    the :meth:`step` method inputs a ``loss`` tensor to be differentiated,
+    and backpropagation is triggered one or more times inside the optimizer.
+
+    Derived classes must implement :meth:`step` to compute derivatives and
+    update parameters in-place.
+
+    Example::
+
+        tr = poutine.trace(model).get_trace(*args, **kwargs)
+        loss = -tr.log_prob()
+        params = {name: pyro.param(name).unconstrained()
+                  for name in pyro.get_param_store().get_all_param_names()}
+        optim.step(loss, params)
+    """
+    def step(self, loss, params):
+        """
+        Performs an optimization step on parameters given a differentiable
+        ``loss`` tensor.
+
+        :param torch.Tensor loss: A differentiable tensor to be minimized.
+            Some optimizers require this to be differentiable multiple
+            times.
+        :param dict params: A dictionary mapping param name to unconstrained
+            value as stored in the param store.
+        """
+        raise NotImplementedError
+
+
+class PyroMultiOptimizer(MultiOptimizer):
+    """
+    Facade to wrap :class:`~pyro.optim.optim.PyroOptim` objects
+    in a :class:`MultiOptimizer` interface.
+    """
+    def __init__(self, optim):
+        self.optim = optim
+
+    def step(self, loss, params):
+        names = []
```

Comment:
I guess we don't need to store this list in this method.

Reply:
So I propose we use `params` as a list (not a dict) of parameters. :)

Reply:
After discussion with @eb8680 we think it would be cleaner to start using dictionaries rather than lists in `PyroOptim` in Pyro 0.3. I also find it cleaner to use dicts rather than lists when combining optimizers in `MixedMultiOptimizer`. Does this sound reasonable?